c-scape/src/threads/key.rs

```diff
@@ -0,0 +1,121 @@
+use alloc::boxed::Box;
+use core::cell::Cell;
+use core::ptr::null_mut;
+use errno::{set_errno, Errno};
+use libc::{c_int, c_void};
+
+use origin::sync::Mutex;
+
+const PTHREAD_KEYS_MAX: u32 = 128;
+const PTHREAD_DESTRUCTOR_ITERATIONS: u8 = 4;
+
+#[derive(Clone, Copy)]
+struct KeyData {
+    current_key: libc::pthread_key_t,
+    destructors: [Option<unsafe extern "C" fn(_: *mut c_void)>; PTHREAD_KEYS_MAX as usize],
+}
+
+static KEY_DATA: Mutex<KeyData> = Mutex::new(KeyData {
+    current_key: 0,
+    destructors: [None; PTHREAD_KEYS_MAX as usize],
+});
+
+const VALUES_INIT: Cell<*mut c_void> = Cell::new(null_mut());
+
+#[thread_local]
+static VALUES: [Cell<*mut c_void>; PTHREAD_KEYS_MAX as usize] =
+    [VALUES_INIT; PTHREAD_KEYS_MAX as usize];
+
+#[thread_local]
+static HAS_REGISTERED_CLEANUP: Cell<bool> = Cell::new(false);
+
+#[no_mangle]
+unsafe extern "C" fn pthread_getspecific(key: libc::pthread_key_t) -> *mut c_void {
+    libc!(libc::pthread_getspecific(key));
+
+    VALUES[key as usize].get()
```

Comment:
The current `pthread_key_create` and `pthread_key_delete` don't clear the value for all threads. So here's where I realize that what I said earlier won't work out as I'd hoped :-}.

What if we had a shared `[AtomicU64; PTHREAD_KEYS_MAX]` that kept an epoch counter for each slot, `VALUES` held a `(*mut c_void, u64)` of the value and the current value of the counter. That way, `pthread_key_delete` could increment the shared counter, and `pthread_getspecific` could check whether the thread-local counter in the slot matches the shared counter value?

Another option is to add something to origin to store this data inside the `ThreadData`, and teach it how to clear out a slot across all threads. I hesitate to go there though, because origin is ideally focusing on features needed by Rust programs.

Reply:
In hindsight I should've realized one should be able to reuse deleted keys...

Let's do epochs, we make `pthread_key` a tuple of epoch number and index number. `KeyData` also gains the aforementioned array (or maybe just a single atomic?). Then we simply need to compare the key's generation number with that of the thread-local value.

If a key's epoch is lower than the thread-local we error, if it's higher we set it to null before doing anything, and basically just lazily treat it as null.